crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+            if let Some(throttle_write_iops_device) = blkio.throttle_write_iops_device() {
+                let mut file = OpenOptions::new()
+                    .write(true)
+                    .open(controller_path.join("blkio.throttle.write_iops_device"))?;
+                for device in throttle_write_iops_device {
+                    file.write_all(
+                        format!("{}:{} {}", device.major(), device.minor(), device.rate())
+                            .as_bytes(),
+                    )?;
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl CgroupV1 {
+    pub fn new(base: PathBuf, path: PathBuf) -> Self {
+        CgroupV1 {
+            path: path,
+            base: base,
+        }
+    }
+
+    fn get_controller(&self, controller: &str) -> Result<PathBuf> {
+        let controller_path = self.base.join(controller);
+        if !controller_path.exists() {
+            return Err(Error::Others(format!(
+                "controller not found {}: {}",
+                controller,
+                controller_path.to_str().unwrap(),
+            )));
+        }
+
+        let p = controller_path.join(&self.path);
+        if !p.exists() {
+            std::fs::create_dir_all(&p)?;
+        }
+        Ok(p)
+    }
+}
+
+pub struct CgroupV2 {
+    base: PathBuf,
+    path: PathBuf,
+}
+
+impl CgroupV2 {
+    pub fn new(base: PathBuf, path: PathBuf) -> Self {
+        CgroupV2 {
+            base: base,
+            path: path,
+        }
+    }
+
+    fn get_file(&self, name: &str) -> PathBuf {
+        self.base.join(&self.path).join(name)
+    }
+}
+
+fn ensure_write_file(path: std::path::PathBuf, content: &str) -> Result<()> {
+    let parent = path.parent().unwrap();
+    if !parent.exists() {
+        create_dir_all(parent).map_err(|e| {
+            Error::Others(format!(
+                "could not create parent cgroup dir {}: {}",
+                parent.to_str().unwrap(),
+                e
+            ))
+        })?;
+    }
+    let mut file = OpenOptions::new().write(true).open(&path).map_err(|e| {
+        Error::Others(format!(
+            "could not open cgroup file {}: {}",
+            path.to_str().unwrap(),
+            e
+        ))
+    })?;
+    file.write_all(content.as_bytes())?;
+    Ok(())
+}
+
+impl Cgroup for CgroupV2 {
+    fn version(&self) -> Version {
+        Version::V2
+    }
+
+    fn open(&self) -> Result<RawFD> {
+        let path = self.base.join(&self.path);
+        create_dir_all(&path)?;
+        nix::fcntl::open(
+            path.as_path(),
+            nix::fcntl::OFlag::O_DIRECTORY
+                | nix::fcntl::OFlag::O_CLOEXEC
+                | nix::fcntl::OFlag::O_RDONLY,
+            nix::sys::stat::Mode::empty(),
+        )
+        .map_err(|e| std::io::Error::from(e).into())
+    }
+
+    fn delete(&self) -> Result<()> {
+        let path = self.base.join(&self.path);
+        if path.exists() {
+            std::fs::remove_dir(path)?;
```

Comment:
The correct way to delete a cgroup is to use rmdir (you can test this out in a terminal).
For instance if you try to `rm -rf` a cgroup dir (which will try to recursively delete every object in the dir) you will get an error because cgroups are a special filesystem and it doesn't make sense to delete files inside a cgroup. If the cgroup exists the files must also exist, so `rmdir` is how to delete the group.